### PR TITLE
Processing metadata tools

### PR DESCRIPTION
**Pull request by alexbruy**
## Description

Add a few more native algorithms to deal with layer metadata:

- Add history metadata to add a new history entry to existing entries defined for a layer
- Update layer metadata to copy non-empty metadata fields from input layer to a target layer and keeping empty fields from input unchanged in the target
- Set metadata fields to set simple metadata fields (identifier, parent identifier, title, type, encoding, language, crs, abstract, fees)

Followup #59411.

### CODE REVIEW

File: src/analysis/processing/qgsmetadataalgorithms.cpp [hunk 1]
```diff
@@ -248,4 +248,311 @@ QVariantMap QgsExportLayerMetadataAlgorithm::processAlgorithm( const QVariantMap
...
+}
+
+QString QgsUpdateLayerMetadataAlgorithm::groupId() const
+{
+  return QStringLiteral( "metadatatools" );
+}
+
+QString QgsUpdateLayerMetadataAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Copies all non-empty metadata fields from an input layer to a target layer.\n\nLeaves empty input fields unchaged in the target." );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsmetadataalgorithms.cpp:351]:
```suggestion
  return QObject::tr( "Copies all non-empty metadata fields from an source layer to a target layer.\n\nLeaves empty input fields unchanged in the target." );
```

File: src/analysis/processing/qgsmetadataalgorithms.cpp [hunk 2]
```diff
@@ -248,4 +248,311 @@ QVariantMap QgsExportLayerMetadataAlgorithm::processAlgorithm( const QVariantMap
...
+}
+
+QgsUpdateLayerMetadataAlgorithm *QgsUpdateLayerMetadataAlgorithm::createInstance() const
+{
+  return new QgsUpdateLayerMetadataAlgorithm();
+}
+
+void QgsUpdateLayerMetadataAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterMapLayer( QStringLiteral( "INPUT" ), QObject::tr( "Source layer" ) ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsmetadataalgorithms.cpp:361]: I'd probably go with "SOURCE" here, as "INPUT" is a bit ambiguous (either layer could reasonably be assumed to be the input here)

```suggestion
  addParameter( new QgsProcessingParameterMapLayer( QStringLiteral( "SOURCE" ), QObject::tr( "Source layer" ) ) );
```

File: src/analysis/processing/qgsmetadataalgorithms.cpp [hunk 3]
```diff
@@ -248,4 +248,311 @@ QVariantMap QgsExportLayerMetadataAlgorithm::processAlgorithm( const QVariantMap
...
+  addParameter( new QgsProcessingParameterMapLayer( QStringLiteral( "TARGET" ), QObject::tr( "Target layer" ) ) );
+  addOutput( new QgsProcessingOutputMapLayer( QStringLiteral( "OUTPUT" ), QObject::tr( "Updated layer" ) ) );
+}
+
+bool QgsUpdateLayerMetadataAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsMapLayer *inputLayer = parameterAsLayer( parameters, QStringLiteral( "INPUT" ), context );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsmetadataalgorithms.cpp:368]:
```suggestion
  QgsMapLayer *inputLayer = parameterAsLayer( parameters, QStringLiteral( "SOURCE" ), context );
```

File: src/analysis/processing/qgsmetadataalgorithms.cpp [hunk 4]
```diff
@@ -248,4 +248,311 @@ QVariantMap QgsExportLayerMetadataAlgorithm::processAlgorithm( const QVariantMap
...
+  QgsMapLayer *targetLayer = parameterAsLayer( parameters, QStringLiteral( "TARGET" ), context );
+
+  if ( !inputLayer )
+    throw QgsProcessingException( QObject::tr( "Invalid input layer" ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsmetadataalgorithms.cpp:372]:
```suggestion
    throw QgsProcessingException( QObject::tr( "Invalid source layer" ) );
```

**Comment by nyalldawson** [comment]
Awesome!! :smile:

File: src/analysis/processing/qgsmetadataalgorithms.cpp (hunk 1 above)
Comment (alexbruy) [reply src/analysis/processing/qgsmetadataalgorithms.cpp]: Fixed and used the same approach in Copy Metadata algorithm

File: src/analysis/processing/qgsmetadataalgorithms.cpp (hunk 2 above)
Comment (alexbruy) [reply src/analysis/processing/qgsmetadataalgorithms.cpp]: Fixed and used the same approach in Copy Metadata algorithm

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**